docs: move inappropriate content out of /docs/Concepts and into new sections; expand /docs/Concepts (#159)
spokestack/spokestack-website: content/docs/Concepts/speech-pipeline.md

---
title: Speech Pipeline
navId: Speech Pipeline (Concepts)
description: Understanding the Spokestack Speech Pipeline
draft: false
tags: Speech Pipeline
---

> The speech pipeline is the main way you interact with Spokestack’s VAD, wake word, and speech recognition. The speech pipeline is an extensible audio processing pipeline that includes a variety of built-in speech processors for Voice Activity Detection (VAD), wake word activation, and Automatic Speech Recognition (ASR).

Reviewer: ```suggestion
The speech pipeline is the main way you interact with Spokestack’s voice activity detection (VAD), wake word detection, and automatic speech recognition (ASR). The speech pipeline is an extensible audio processing pipeline that includes a variety of built-in speech processors for all the above stages.
```
Do we want to link to the concept pages here too?

> This pipeline seamlessly integrates VAD-triggered wake word detection using on-device machine learning models with transcribing utterances using ASR. It runs as a [soft real-time](https://en.wikipedia.org/wiki/Real-time_computing#Criteria_for_real-time_computing) system and its components must be as responsive as possible.

Reviewer: ```suggestion
This pipeline seamlessly integrates VAD-triggered wake word detection using on-device machine learning models with speech transcription. It runs as a [soft real-time](https://en.wikipedia.org/wiki/Real-time_computing#Criteria_for_real-time_computing) system, and its components must be as responsive as possible.
```

> This pipeline seamlessly integrates VAD-triggered wake word detection using on-device machine learning models with speech transcription. It runs as a [soft real-time](https://en.wikipedia.org/wiki/Real-time_computing#Criteria_for_real-time_computing) system, and its components must be as responsive as possible.

Reviewer: ```suggestion
This pipeline seamlessly integrates VAD-triggered wake word detection using on-device machine learning models with speech transcription. It runs as a [soft real-time](https://en.wikipedia.org/wiki/Real-time_computing#Criteria_for_real-time_computing) system.
```
I know I wrote this, but I'm having second thoughts—including "as responsive as possible" is both hand-wavey and more of a guideline for developing a new pipeline stage; I think we can just cut the sentence off earlier.

> [Voice activity detection](/docs/Concepts/vad) enables the pipeline to listen to small segments of audio and determine if speech is present. To keep computation usage low for edge devices, the rest of the pipeline does not proceed if the VAD does not detect speech.

Reviewer 2: ```suggestion
[Voice activity detection](/docs/concepts/vad) enables the pipeline to listen to small segments of audio and determine if speech is present. To keep computation usage low for edge devices, the rest of the pipeline does not proceed if the VAD does not detect speech.
```

> [A wake word](/docs/Concepts/wake-word) enables the pipeline to listen to speech audio and determine if a name from a set of recognized names is spoken. This fulfills two objectives: one, only listen in on a conversation if the wake word is being spoken, and two preserve the higher-cost ASR component to transcribe only when the user wants to talk.

Reviewer 2: ```suggestion
[A wake word](/docs/concepts/wake-word) enables the pipeline to listen to speech audio and determine if a name from a set of recognized names is spoken. This fulfills two objectives: one, only listen in on a conversation if the wake word is being spoken, and two preserve the higher-cost ASR component to transcribe only when the user wants to talk.
```

> [A wake word](/docs/Concepts/wake-word) enables the pipeline to listen to speech audio and determine if a name from a set of recognized names is spoken. This fulfills two objectives: one, only listen in on a conversation if the wake word is being spoken, and two preserve the higher-cost ASR component to transcribe only when the user wants to talk.

Reviewer: ```suggestion
[A wake word](/docs/concepts/wake-word) enables the pipeline to listen to speech audio and determine if a name from a set of recognized names is spoken. This fulfills two objectives: one, only listen in on a conversation if the wake word is being spoken; and two, ensure the higher-cost ASR component to runs only when the user wants to be understood by an app.
```

> When Spokestack detects a wake word, the speech pipeline begins transcribing the user's speech until they stop talking for a pre-set amount of time, or a total activation time limit elapses. The technology for converting spoken words to text is known as [automatic speech recognition (ASR)](/docs/Concepts/asr).

Reviewer 2: ```suggestion
When Spokestack detects a wake word, the speech pipeline begins transcribing the user's speech until they stop talking for a pre-set amount of time, or a total activation time limit elapses. The technology for converting spoken words to text is known as [automatic speech recognition (ASR)](/docs/concepts/asr).
```

> When Spokestack detects a wake word, the speech pipeline begins transcribing the user's speech until they stop talking for a pre-set amount of time, or a total activation time limit elapses. The technology for converting spoken words to text is known as [automatic speech recognition (ASR)](/docs/Concepts/asr).

Reviewer: ```suggestion
When Spokestack detects a wake word, the speech pipeline begins transcribing the user's speech until they stop talking for a pre-set amount of time, or a total activation time limit elapses. The technology for converting spoken words to text is known as [automatic speech recognition](/docs/concepts/asr).
```
We don't need to reintroduce the initialism, but using the full phrase makes sense here.

> When Spokestack detects a wake word, the speech pipeline begins transcribing the user's speech until they stop talking for a pre-set amount of time, or a total activation time limit elapses. The technology for converting spoken words to text is known as [automatic speech recognition (ASR)](/docs/concepts/asr).

Reviewer: ```suggestion
When Spokestack detects a wake word, the speech pipeline begins transcribing the user's speech until they stop talking for a pre-set amount of time, or a total activation time limit elapses. The technology for converting spoken words to text is known as [automatic speech recognition](/docs/concepts/asr).
```
We've already introduced the initialism, but it does make sense to use the full form here.

> The speech pipeline is the first piece of the puzzle in any voice interaction and is responsible for capturing user audio and translating it into text. Configuring it entails choices about whether or not to use a wakeword to activate ASR, what kind of preprocessing to perform on audio before sending it to ASR, and which ASR service to use. These choices can all be made [individually or through the use of configuration profiles](/docs/Machine Learning/pipeline-configuration).

Reviewer 2: ```suggestion
The speech pipeline is the first piece of the puzzle in any voice interaction and is responsible for capturing user audio and translating it into text. Configuring it entails choices about whether or not to use a wakeword to activate ASR, what kind of preprocessing to perform on audio before sending it to ASR, and which ASR service to use. These choices can all be made [individually or through the use of configuration profiles](/docs/machine-learning/pipeline-configuration).
```

> The speech pipeline is the first piece of the puzzle in any voice interaction and is responsible for capturing user audio and translating it into text. Configuring it entails choices about whether or not to use a wakeword to activate ASR, what kind of preprocessing to perform on audio before sending it to ASR, and which ASR service to use. These choices can all be made [individually or through the use of configuration profiles](/docs/Machine Learning/pipeline-configuration).

Reviewer: ```suggestion
The speech pipeline is the first piece of the puzzle in any voice interaction and is responsible for capturing user audio and translating it into text. Configuring it entails choices about whether or not to use a wake word to activate ASR, what kind of preprocessing to perform on audio before sending it to ASR, and which ASR service to use. These choices can all be made [individually or through the use of configuration profiles](/docs/machine-learning/pipeline-configuration).
```
Starting a final paragraph with "the first piece of the puzzle" without further elaboration on that feels a little odd to me.

> The speech pipeline is the first piece of the puzzle in any voice interaction and is responsible for capturing user audio and translating it into text. Configuring it entails choices about whether or not to use a wakeword to activate ASR, what kind of preprocessing to perform on audio before sending it to ASR, and which ASR service to use. These choices can all be made [individually or through the use of configuration profiles](/docs/machine-learning/pipeline-configuration).

Reviewer: Starting a final paragraph with "the first piece of the puzzle" without any further talk of the puzzle seems odd to me.
